sign/dilithium/mode3/internal/sample.go

```diff
@@ -2,9 +2,192 @@ package internal
+// sampling variants like PolyDeriveUniformX4.
+var DeriveX4Available = f1600x4.Available && !UseAES
+
+// For each i, sample ps[i] uniformly with coefficients of norm less than γ₁
+// using the the given seed and nonces[i].  ps[i] may be nil and is ignored
+// in that case.  ps[i] will not be normalized, but have coefficients in the
+// interval (q-γ₁,q+γ₁).
+//
+// Can only be called when DeriveX4Available is true.
+func PolyDeriveUniformLeGamma1X4(ps [4]*common.Poly, seed *[48]byte,
+	nonces [4]uint16) {
+	var perm f1600x4.State
+	state := perm.Initialize()
+
+	// Absorb the seed in the four states
+	for i := 0; i < 6; i++ {
+		v := binary.LittleEndian.Uint64(seed[8*i : 8*(i+1)])
+		for j := 0; j < 4; j++ {
+			state[i*4+j] = v
+		}
+	}
+
+	// Absorb the nonces, the SHAKE256 domain separator (0b1111), the
+	// start of the padding (0b...001) and the end of the padding 0b100...
+	// Recall that the rate of SHAKE256 is 136 --- i.e. 17 uint64s.
+	for j := 0; j < 4; j++ {
+		state[6*4+j] = uint64(nonces[j]) | (0x1f << 16)
+		state[16*4+j] = 0x80 << 56
+	}
+
+	var idx [4]int // indices into ps
+	for j := 0; j < 4; j++ {
+		if ps[j] == nil {
+			idx[j] = 256 // mark nil polynomial as completed
+		}
+	}
+
+	// Each try requires 15 bits.  15 does not divide into 64, nor in 136,
+	// so we will have to carry some bits from a previous uint64 to the next.
+	var carry [4]uint32
+
+	// Shift is the amount of bits in the carry.
+	var shift [4]uint
+
+	for {
+		// Applies KeccaK-f[1600] to state to get the next 17 uint64s of each
+		// of the four SHAKE256 streams.
+		perm.Permute()
+
+		done := true
+
+		for j := 0; j < 4; j++ {
+			if idx[j] == 256 {
+				continue
+			}
+			done = false
+
+			for i := 0; i < 17; i++ {
+				var t [4]uint32
+				tCount := 3
+
+				// Get the next three or four 20 bit numbers.
+				qw := state[i*4+j]
+				qwl := (qw << shift[j]) | uint64(carry[j])
+				t[0] = uint32(qwl & 0xfffff)
+				t[1] = uint32((qwl >> 20) & 0xfffff)
+				t[2] = uint32((qwl >> 40) & 0xfffff)
+
+				if shift[j] == 16 {
+					t[3] = uint32(qw >> 44)
+					shift[j] = 0
+					carry[j] = 0
+					tCount = 4
+				} else {
+					shift[j] += 4
+					carry[j] = uint32(qw >> (64 - shift[j]))
+				}
+
+				// Check if they're coefficients.
+				for k := 0; k < tCount; k++ {
+					if t[k] <= 2*common.Gamma1-2 {
+						ps[j][idx[j]] = common.Q + common.Gamma1 - 1 - t[k]
+						idx[j]++
+						if idx[j] == 256 {
+							break
+						}
+					}
+				}
+
+				if idx[j] == 256 {
+					break
+				}
+			}
+		}
+
+		if done {
+			break
+		}
+	}
+}
+
+// For each i, sample ps[i] uniformly from the given seed and nonces[i].
+// ps[i] may be nil and is ignored in that case.
+//
+// Can only be called when DeriveX4Available is true.
+func PolyDeriveUniformX4(ps [4]*common.Poly, seed *[32]byte, nonces [4]uint16) {
+	var perm f1600x4.State
+	state := perm.Initialize()
+
+	// Absorb the seed in the four states
```

Comment:
This code looks to be repeated in the previous function, maybe it will go into a small function.

Reply:
It's slightly different: the bound on `i` is 4 in one and 6 in the other